cpp/src/svm/kernelcache.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_utils.h>
+#include <linalg/gemm.h>
+#include "cache/cache.h"
+#include "common/cumlHandle.hpp"
+#include "common/host_buffer.hpp"
+#include "gram/grammatrix.h"
+#include "ml_utils.h"
+
+namespace ML {
+namespace SVM {
+
+/**
+ * @brief Collect rows of the training data into contiguous space
+ *
+ * The working set is a subset of all the training examples. Here we collect
+ * all the training vectors that are in the working set.
+ *
+ * @param [in] x training data in column major format, size [n_rows x n_cols]
+ * @param [in] n_rows
+ * @param [in] n_cols
+ * @param [out] x_ws training vectors in the working set in column major format, size [n_ws x n_cols]
+ * @param [in] n_ws the number of elements in the working set
+ * @param [in] ws_idx working set indices (row indices of x), size [n_ws]
+ */
+template <typename math_t>
+__global__ void get_rows(const math_t *x, int n_rows, int n_cols, math_t *x_ws,
```

Comment:
There are similar functions to get_rows in matrix.h file. Please check the ones in here https://github.com/rapidsai/cuml/blob/branch-0.10/cpp/src_prims/matrix/matrix.h

I know copyRows function in matrix.h uses thrust but I'd recommend either update this existing function or move your function to this file. You can overload this function as well.